parser-typechecker/src/Unison/Hashing/V2/BuildHashable.hs

```diff
@@ -0,0 +1,140 @@
+module Unison.Hashing.V2.BuildHashable where
+
+import qualified Crypto.Hash as CH
+import qualified Data.ByteArray as BA
+import qualified Data.ByteString as B
+import Data.ByteString.Builder (doubleBE, int64BE, toLazyByteString, word64BE)
+import qualified Data.ByteString.Lazy as BL
+import qualified Data.Map as Map
+import qualified Data.Set as Set
+import U.Util.Hash (Hash)
+import qualified U.Util.Hash as Hash
+import Unison.Prelude
+import Unison.Util.Relation (Relation)
+import qualified Unison.Util.Relation as Relation
+import Unison.Util.Relation3 (Relation3)
+import qualified Unison.Util.Relation3 as Relation3
+import Unison.Util.Relation4 (Relation4)
+import qualified Unison.Util.Relation4 as Relation4
+
+data Token h
+  = Tag !Word8
+  | Bytes !ByteString
+  | Int !Int64
+  | Text !Text
+  | Double !Double
+  | Hashed !h
+  | Nat !Word64
+
+class Accumulate h where
+  accumulate :: [Token h] -> h
+  fromBytes :: ByteString -> h
+  toBytes :: h -> ByteString
+
+accumulateToken :: (Accumulate h, Hashable t) => t -> Token h
+accumulateToken = Hashed . accumulate'
+
+hash, accumulate' :: (Accumulate h, Hashable t) => t -> h
+hash = accumulate'
+accumulate' = accumulate . (hashVersion :) . tokens
+  where
+    hashVersion = Tag 2
+
+class Hashable t where
+  tokens :: Accumulate h => t -> [Token h]
+
+instance Hashable a => Hashable [a] where
+  tokens = map accumulateToken
+
+instance (Hashable a, Hashable b) => Hashable (a, b) where
+  tokens (a, b) = [accumulateToken a, accumulateToken b]
```

Comment:
I notice we have a lot of overlap between our Hashables; e.g. AFAICT all of these would have the same hash:

```
[1, 2]
(1, 2)
Set.fromList [1, 2]
```
etc.;

Should we be tagging these somehow? Or maybe prepending a `Text "Set"` to the front or something to disambiguate?

Seems like a really subtle bug just waiting to happen 😅